sql/analyzer/indexed_joins.go

```diff
@@ -0,0 +1,633 @@
+	// implementation (which can be faster for tables that fit into memory). Over time, we should join these two
+	// implementations.
+	if !replacedTableWithIndexedAccess {
+		return n, nil
+	}
+
+	return withIndexedTableAccess, nil
+}
+
+// replaceTableAccessWithIndexedAccess replaces table access with indexed access where possible. This can't be a
+// standard bottom-up transformation, because we need information that isn't accessible in the node itself or in the
+// parent. Specifically, the available schema to right-hand branches of the tree is constructed at runtime as the
+// concatenation of the parent row (passed into row.Iter()) and the row returned by the left-hand branch of the join.
+// This is basically an in-order concatenation of columns in all tables to the left of the one being examined, including
+// from the left branches of parent nodes, which means there is no way to construct it given just the parent node.
+func replaceTableAccessWithIndexedAccess(
+		node sql.Node,
+		schema sql.Schema,
+		scope *Scope,
+		joinIndexes joinIndexesByTable,
+		exprAliases ExprAliases,
+		tableAliases TableAliases,
+) (sql.Node, error) {
+	switch node := node.(type) {
+	case *plan.TableAlias, *plan.ResolvedTable:
+		// If the available schema makes an index on this table possible, use it, replacing the table with indexed access
+		indexes := joinIndexes[node.(sql.Nameable).Name()]
+		indexToApply := indexes.getUsableIndex(schema)
+		if indexToApply == nil {
+			return node, nil
+		}
+
+		return plan.TransformUp(node, func(node sql.Node) (sql.Node, error) {
+			switch node := node.(type) {
+			case *plan.ResolvedTable:
+				if _, ok := node.Table.(sql.IndexAddressableTable); !ok {
+					return node, nil
+				}
+
+				keyExprs := createIndexLookupKeyExpression(indexToApply, exprAliases, tableAliases)
+				keyExprs, err := FixFieldIndexesOnExpressions(scope, schema, keyExprs...)
+				if err != nil {
+					return nil, err
+				}
+
+				return plan.NewIndexedTableAccess(node, indexToApply.index, keyExprs), nil
+			default:
+				return node, nil
+			}
+		})
+	case *plan.IndexedJoin:
+		// Recurse the down the left side with the input schema
+		left, err := replaceTableAccessWithIndexedAccess(node.LeftBranch(), schema, scope, joinIndexes, exprAliases, tableAliases)
+		if err != nil {
+			return nil, err
+		}
+
+		// then the right side, appending the schema from the left
+		right, err := replaceTableAccessWithIndexedAccess(node.RightBranch(), append(schema, left.Schema()...), scope, joinIndexes, exprAliases, tableAliases)
+		if err != nil {
+			return nil, err
+		}
+
+		// the condition's field indexes might need adjusting if the order of tables changed
+		cond, err := FixFieldIndexes(scope, append(schema, append(left.Schema(), right.Schema()...)...), node.Cond)
+		if err != nil {
+			return nil, err
+		}
+
+		return plan.NewIndexedJoin(left, right, node.JoinType(), cond), nil
+	case *plan.Sort:
```

Comment:
This isn't pushing them down... this is examining the entire tree to determine the schema of columns that each IndexedJoin node will get in its RowIter() calls. Has to be done this way (looking at the entire tree top down) since that's how these context rows are built at run time. I can't only examine the tree starting with the top join node, since some outer wrapping nodes (like a CrossJoin or a Union) will contribute columns to this schema. Most won't, which is why there is so much boilerplate. I did a refactoring to reduce the repetition here, but it's still not great.

I'll think some more about how to formalize this, since it's kind of confusing and is bound to come up again.